File: src/main/java/ie/just/another/java/repo/NestedParentheses.java
package ie.just.another.java.repo;

public class NestedParentheses {

    private NestedParentheses() {

    }

    /**
     * Checks it the round parentheses in a string are properly nested
     * eg:
     *    "()" or "(bc(d)e)" parentheses are properly nested
     *    ")(" or ")a(ad)e(" parentheses are not properly nested
     *
     * @param s the string to check
     * @return a boolean stating if the parentheses are properly nested
     */
    public static boolean isCorrectlyNested(String s) {

        int validParenthesesIndex = 0;
        for(int i=0;i<s.length();i++){
            char parentheses = s.charAt(i);
            if('(' == parentheses){
                validParenthesesIndex = validParenthesesIndex + 1;
            } else if (')' == parentheses) {
                validParenthesesIndex = validParenthesesIndex - 1;
            } else {
                System.out.println("This is not a parentheses");
            }
            if(validParenthesesIndex < 0) {
                return false;
            }
        }
        return true;
    }

}

File: src/main/scala/ie/just/another/scala/repo/NestedParentheses.scala
package ie.just.another.scala.repo

import scala.annotation.tailrec

object NestedParentheses {

  /**
   * Checks it the round parentheses in a string are properly nested
   * eg:
   * "()" or "(bc(d)e)" parentheses are properly nested
   * ")(" or ")a(ad)e(" parentheses are not properly nested
   *
   * @param s the string to check
   * @return a boolean stating if the parentheses are properly nested
   */
  def isCorrectlyNested(s: String): Boolean = {

    @tailrec
    def check(chars: Array[Char], index: Int = 0): Int = {
      if (chars.isEmpty || index < 0) {
        index
      } else if (chars.head == '(') {
        check(chars.tail, index + 1)
      } else if (chars.head == ')') {
        check(chars.tail, index - 1)
      } else {
        check(chars.tail, index)
      }
    }

    if (check(s.toCharArray) == 0) true else false
  }

}
